Ruby xDS interop test client

stanley-cheung opened the pull request:
This PR implements the xDS interop client in Ruby.

Current state:
- All green adhoc build:
   - https://g3c.corp.google.com/results/invocations/9a67d82c-915c-4606-9fd9-ff896ae56d9d/targets

stanley-cheung commented:
@apolcyn @ericgribkoff Just got an all green run (see above), so the code should be ready to be reviewed. Thanks!

stanley-cheung commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,193 @@
+    # wait until all RPCs are done, or we are past timeout
+    while ($watchers[index]['rpcs_needed'] > 0) &&
+          (Time.now.to_f < finish_time)
+      sleep(0.1)
```
I am not sure if there's a better mechanism to do this in Ruby. Open to ideas.

stanley-cheung replied:
This is replaced by a `ConditionVariable` below

stanley-cheung commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,193 @@
+require_relative '../src/proto/grpc/testing/messages_pb'
+require_relative '../src/proto/grpc/testing/test_services_pb'
+
+# Some global variables to be shared by server and client
```
Not sure if this is OK - or should I make some sort of class to wrap these?

apolcyn replied:
current approach LGTM

apolcyn commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,193 @@
+# Some global variables to be shared by server and client
+$watchers = Array.new
+$watchers_mutex = Mutex.new
+$current_request_id = 0;
```
I think it would be easier to reason about the thread safety of the logic regarding `current_request_id` if it was also protected under `watchers_mutex`

Can we change the logic so that we only read or write to `current_request_id` while holding `watchers_mutex` ?

To be clear that this is protected by `watchers_mutex`, can we rename to `watchers_ current_request_id` ?

stanley-cheung replied:
After removing `start_id` and `end_id` below, there's no need for `current_request_id` anymore. So I removed that all together.

apolcyn commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,193 @@
+      index = $watchers.length
+      # Add a watcher for this range of id's
+      $watchers << {
+        "start_id" => start_id,
```
nit: we can remove `start_id` and `end_id` variables, and set "start_id" directly to `$current_request_id + 1`, and "end_id" directly to `$current_request_id + 1 + req['num_rpcs']`

Related to above comment about only synchronizing `current_request_id`

stanley-cheung replied:
Fixed

apolcyn commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,193 @@
+  def get_client_stats(req, _call)
+    start_id = $current_request_id + 1
+    end_id = start_id + req['num_rpcs'] # watch the next X rpcs
+    start_time = Time.now.to_f
```
nit: let's use monotonic time, can we set `start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)` ?

stanley-cheung replied:
Fixed.

apolcyn commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,193 @@
+    finish_time = start_time + req['timeout_sec']
+    index = 0
+    $watchers_mutex.synchronize do
+      index = $watchers.length
```
nit: rather than using an `index` into `$watchers`, can we create local variable `watcher = { ... }` and then add `watcher` to `$watchers` ?

stanley-cheung replied:
Fixed

apolcyn commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,193 @@
+      }
+    end
+    # wait until all RPCs are done, or we are past timeout
+    while ($watchers[index]['rpcs_needed'] > 0) &&
```
we can use a `ConditionVariable` to replace the sleeps here.

Let's add a new global, `$watchers_cv = ConditionVariable.new`

Then do something like this (this also removes `watcher` after it's no longer used)

```
$watchers_mutex.synchronize do
  watcher = { ... }
  $watchers << watcher
  seconds_remaining = Process.clock_gettime(Process::CLOCK_MONOTONIC) - finish_time
  while watcher['rpcs_needed'] > 0 && seconds_remaining > 0
    $watchers_cv.wait($watchers_mutex, seconds_remaining)
    seconds_remaining = Process.clock_gettime(Process::CLOCK_MONOTONIC) - finish_time
  end
  $watchers.delete_at($watchers.index watcher)
end
LoadBalancerStatsResponse.new(...)
```

stanley-cheung replied:
Fixed. Thanks for pointing me to `ConditionVariable`.

apolcyn commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,193 @@
+            watcher['rpcs_by_peer'][remote_peer] += 1
+          end
+        end
+      end
```
related to above change, we'll need to broadcast to `watchers_cv` down here

stanley-cheung replied:
Fixed

apolcyn commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,193 @@
+    $current_request_id += 1
+    $watchers_mutex.synchronize do
+      $watchers.each do |watcher|
+        if (watcher['start_id'] <= $current_request_id) &&
```
nit: if we remove `watchers` from the `$watchers` list after we're done with them, then I think there's no need to have the `end_id` field within watcher objects at all.

I think we can delete this conditional logic then and unconditionally run the body of this if statement

stanley-cheung replied:
Yes this works. Basically the idea is that when we add a `watcher`, we are interested in the _next_ `num_rpcs`. So each subsequent completed rpcs will just decrement every `watcher`'s `num_rpcs` in the list by 1. When a `watcher` reaches `num_rpcs = 0` (or timed out), it's removed from the `watchers` list. So it can be reasoned that any successful rpc is useful to _every_ `watcher` currently still in the list.

apolcyn commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,193 @@
+  # The client just sends unary rpcs continously in a regular interval
+  stub = create_stub(opts)
+  duration_per_query = (1.0 / opts['qps'].to_f)
+  client_thread = Thread.new {
```
per the spec I think we need to add an `--num_channels` option here, and then create `--num_channels` stubs, eaching having their own one of these threads

stanley-cheung replied:
Added

apolcyn commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,193 @@
+  client_thread = Thread.new {
+    run_test_loop(stub, duration_per_query)
+  }
+
```
let's make sure to join the `client_threads`

I think we could do that by adding a `shutdown` global variable, setting it true here, and checking it at the top of each iteration of the while loop, returning from the loop if shutdown is true.

No need to guard `shutdown` with a mutex since it's a single variable thereby protected by the GIL.

stanley-cheung replied:
The Python test runner is terminating this Ruby client by `subprocess.terminate()`, so a `SIGTERM` signal is being sent to this program. So I modified the server to start with `s.run_till_terminated_or_interrupted(['TERM'])`. And it successfully causes `server_thread.join` to finally return when the Python test runner attempts to terminate this program. After that point, I am able to set the `$shutdown` global variable to cause each of the `client_thread` to terminate and hence they can all `client_thread.join` too. And I verified that we reached the last line of the `main` function properly to shut down.

apolcyn replied:
thanks, this looks good!

stanley-cheung commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,193 @@
+    begin
+      resp = stub.unary_call(req)
+      remote_peer = resp.hostname
+    rescue GRPC::Unavailable => e
```
Fixed

stanley-cheung commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,193 @@
+end
+
+# send 1 rpc every 1/qps second
+def run_test_loop(stub, duration_per_query)
```
Fixed

stanley-cheung commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,193 @@
+  req = SimpleRequest.new()
+
+  while true
+    start = Time.now.to_f
```
Fixed

stanley-cheung commented:
@apolcyn Thanks for the detailed review! I think I have addressed all the comments and am ready for another round of review on your convenience.

The current code passed another all green build: https://g3c.corp.google.com/results/invocations/45cced8e-d10c-4b5b-a7e0-717a731911b4/targets

apolcyn commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,213 @@
+    end
+    LoadBalancerStatsResponse.new(
+      rpcs_by_peer: watcher['rpcs_by_peer'],
+      num_failures: watcher['no_remote_peer'] + watcher['rpcs_needed']
```
Is the correct as per the spec?

I.e., should we just be setting `num_failures: watcher['no_remote_peer']`, rather than also adding `watcher['rpcs_needed']` in `num_failures` ?

cc @ericgribkoff

stanley-cheung replied:
What if `timeout_sec` is hit, but not all the rpcs are sent? I thought that was the intent that those should be counted as `num_failures`. I copied that logic from the c++ xds interop client here: https://github.com/grpc/grpc/blob/master/test/cpp/interop/xds_interop_client.cc#L107. But it would be good if Eric can confirm.

apolcyn replied:
Please let @ericgribkoff confirm here.

FWIW the words from the spec that made me think we shouldn't be added "rpcs_needed" is "If num_rpcs have not completed within timeout_sec, return partial results."

stanley-cheung replied:
@ericgribkoff Yea looks like c# did not add rpcs that haven't completed by timeout. https://github.com/grpc/grpc/pull/22704/files#diff-db0b306f37d8a70b6dc3e0b65896fff4R270. This is inconsistent with the c++ implementation. Please confirm which one is preferred, thanks!

ericgribkoff replied:
This is correct as-is, the intent is that pending rpcs will be added to the failure count as well. I'll update the spec to clarify this.

cc @jtattermusch

stanley-cheung commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,213 @@
+# creates a test stub
+def create_stub(opts)
+  address = "#{opts.server}"
+
```
Fixed

stanley-cheung commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,213 @@
+def run_test_loop(stub, target_seconds_between_rpcs, fail_on_failed_rpcs)
+  include Grpc::Testing
+  req = SimpleRequest.new()
+
```
Fixed. I left a couple in in the `main` method.

stanley-cheung commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,213 @@
+  req = SimpleRequest.new()
+
+  target_next_start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+  while true
```
Fixed

stanley-cheung commented on src/ruby/pb/test/xds_client.rb:
```diff
@@ -0,0 +1,213 @@
+      sleep(sleep_seconds)
+    end
+
+    target_next_start = target_next_start + target_seconds_between_rpcs
```
Fixed

stanley-cheung commented:
@apolcyn I am slightly concerned about this test failure that happened two times in a row now: https://source.cloud.google.com/results/invocations/6b3f125f-3fc1-4720-a634-45b75342b0e4/targets/grpc%2Fcore%2Fpull_request%2Flinux%2Fgrpc_build_artifacts/log

```
++ dirname tools/run_tests/helper_scripts/bundle_install_wrapper.sh
+ cd tools/run_tests/helper_scripts/../../..
++ uname
++ cut -f 1 -d_
+ SYSTEM=Linux
+ '[' Linux == Darwin ']'
+ bundle install
Fetching gem metadata from https://rubygems.org/.........
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
google-protobuf-3.12.0-x86_64-linux requires ruby version >= 2.5, which is
incompatible with the current version, ruby 2.4.1p111
+ export DOCKERHUB_ORGANIZATION=grpctesting
+ DOCKERHUB_ORGANIZATION=grpctesting
+ rake gem:native
rake aborted!
LoadError: cannot load such file -- rake/extensiontask
/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/Rakefile:2:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.4.1@global/gems/rake-12.0.0/exe/rake:27:in `<top (required)>'
(See full trace by running task with --trace)
```

Do you have any idea? It seems like some dependencies have changed.

stanley-cheung commented:
Looks like this is good to go. Let me rebase & squash, remove `grpc_experiment.cfg` and let test run one more time and then I will merge.
